pkg/installmanager/dnscleanup.go

```diff
@@ -0,0 +1,55 @@
+package installmanager
+
+import (
+	awsclient "github.com/openshift/hive/pkg/awsclient"
+
+	log "github.com/sirupsen/logrus"
+
+	awssdk "github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/route53"
+)
+
+// CleanupDNSZone queries the Route53 zone and deletes any A records found. Other record
+// types may be added in the future, but right now this is the only one we're seeing
+// leak and conflict.
+// May no longer be necessary once https://jira.coreos.com/browse/CORS-1195 is fixed.
+func CleanupDNSZone(dnsZoneID, region string, logger log.FieldLogger) error {
+	zoneLogger := logger.WithField("dnsZoneID", dnsZoneID)
+	zoneLogger.Info("cleaning up DNSZone")
+
+	awsClient, err := awsclient.NewClientFromEnvVars(region)
+	if err != nil {
+		return err
+	}
+	recordsOutput, err := awsClient.ListResourceRecordSets(
+		&route53.ListResourceRecordSetsInput{
+			HostedZoneId: awssdk.String(dnsZoneID),
+		},
+	)
+	if err != nil {
+		return err
+	}
+	for _, r := range recordsOutput.ResourceRecordSets {
+		// We're only experiencing problems with A records, so these are all we cleanup for now:
+		if *r.Type == "A" {
+			zoneLogger.WithFields(log.Fields{"name": *r.Name, "type": *r.Type}).Info("deleting A record")
+			request := &route53.ChangeResourceRecordSetsInput{
+				ChangeBatch: &route53.ChangeBatch{
+					Changes: []*route53.Change{
+						{
+							Action:            awssdk.String("DELETE"),
+							ResourceRecordSet: r,
+						},
+					},
+				},
+				HostedZoneId: awssdk.String(dnsZoneID),
+			}
+			_, err := awsClient.ChangeResourceRecordSets(request)
+			if err != nil {
+				return err
```

Comment:
Should this behave like the uninstaller does and loop until there are no failures instead of bailing out when there is a failure?

Reply:
I'm not sure, I think that's effectively what this will do now, we try to cleanup after a failure, and at the start of our next attempt. A failure to delete records here will error out the job and queue it up for another provision attempt. That gets us free backoff and realistically if you're failing to delete a DNS record more than once or twice, it's likely not going to resolve itself and will need manual intervention. In practice I think it's better to let that hit the provision backoff than infinite loop and hammer the API.

Reply:
My biggest concern was that the backoff escalates very quickly. So if there is a failed install followed by a transient error of some kind deleting the A record, then the backoff can quickly get into the hours range.